internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicManagerRepository.java

```diff
@@ -0,0 +1,72 @@
+package com.linkedin.venice.pubsub.manager;
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * A repository of {@link TopicManager} instances, each associated with a specific PubSub region and cluster.
+ * This repository maintains one {@link TopicManager} for each unique PubSub bootstrap server address.
+ * While not mandatory, it is expected that each Venice component will have one and only one instance of this class.
+ */
+public class TopicManagerRepository implements Closeable {
+  private static final Logger LOGGER = LogManager.getLogger(TopicManagerRepository.class);
+  private final Map<String, TopicManager> topicManagers = new VeniceConcurrentHashMap<>();
+  private final TopicManagerContext topicManagerContext;
+  private final String localPubSubAddress;
+
+  public TopicManagerRepository(TopicManagerContext topicManagerContext, String localPubSubAddress) {
+    this.topicManagerContext = Objects.requireNonNull(topicManagerContext, "topicManagerContext cannot be null");
+    this.localPubSubAddress = Objects.requireNonNull(localPubSubAddress, "localPubSubAddress cannot be null");
+  }
+
+  // added in order to help with testing; visibility is package-private for testing purposes
+  TopicManager createTopicManager(String pubSubAddress) {
+    return new TopicManager(pubSubAddress, topicManagerContext);
+  }
+
+  /**
+   * By default, return TopicManager for local PubSub cluster.
+   */
+  public TopicManager getLocalTopicManager() {
+    return getTopicManager(localPubSubAddress);
+  }
+
+  public TopicManager getTopicManager(String pubSubAddress) {
+    return topicManagers.computeIfAbsent(pubSubAddress, k -> createTopicManager(pubSubAddress));
+  }
+
+  public List<TopicManager> getAllTopicManagers() {
+    return new ArrayList<>(topicManagers.values());
```

Comment:
In general, we want to avoid sharing internal data structures between classes (though in the case of tests, that is sometimes useful so we tolerate it, albeit at lower visibility...). In this case, it may be harmless since the collection is immutable, but that is just an implementation, and if it changes later on for some reason, the code may be left in a more fragile state, where some class can inadvertently mutate the internal state of another class through some ill-defined entry point...

The idea of having a public function specifically to invalidate the cache is to offer a precise contract of what can be done with this class, rather than offering lower-level building blocks and letting the caller get creative (which is sometimes good, but also a double-edged sword)...

Anyway, this is a minor nitpick and I leave it up to you.

Reply:
Introducing the invalidateCache API in TopicManagerRepository does reveal a bit about what's going on under the hood of the TopicManagers. But it does seem like an improvement over our current approach. I think I agree with you

Reply:
Replaced `TMR::getAllTopicManagers` with `TMR::invalidateTopicManagerCaches`